contracts/VoteWeighting.sol

```diff
@@ -0,0 +1,487 @@
+    /// @param chainId Chain Id.
+    function addNominee(address nominee, uint256 chainId) external {
+        // Check for the zero address
+        if (nominee == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Check for the chain Id
+        if (chainId == 0) {
+            revert ZeroValue();
+        }
+        else if (chainId > MAX_CHAIN_ID) {
+            revert Overflow(chainId, MAX_CHAIN_ID);
+        }
+
+        // Push a pair of key defining variables into one key
+        // nominee occupies first 160 bits
+        uint256 nomineeChainId = uint256(uint160(nominee));
+        // chain Id occupies no more than next 64 bits
+        nomineeChainId |= chainId << 160;
+
+        // Check for the nominee existence
+        if (mapNomineeIds[nomineeChainId] > 0) {
+            revert NomineeAlreadyExists(nominee, chainId);
+        }
+        mapNomineeIds[nomineeChainId] = setNominees.length;
+        // Push the nominee into the list
+        setNominees.push(nomineeChainId);
+
+        uint256 nextTime = (block.timestamp + WEEK) / WEEK * WEEK;
+        timeWeight[nomineeChainId] = nextTime;
+
+        emit NewNominee(nominee, chainId);
+    }
+
+    /// @dev Checkpoint to fill data common for all nominees.
+    function checkpoint() external {
+        _getSum();
+    }
+
+    /// @dev Checkpoint to fill data for both a specific nominee and common for all nominees.
+    /// @param nominee Address of the nominee.
+    /// @param chainId Chain Id.
+    function checkpointNominee(address nominee, uint256 chainId) external {
+        _getWeight(nominee, chainId);
+        _getSum();
+    }
+
+    /// @dev Get Nominee relative weight (not more than 1.0) normalized to 1e18 (e.g. 1.0 == 1e18).
+    ///         Inflation which will be received by it is inflation_rate * relativeWeight / 1e18.
+    /// @param nominee Address of the nominee.
+    /// @param chainId Chain Id.
+    /// @param time Relative weight at the specified timestamp in the past or present.
+    /// @return weight Value of relative weight normalized to 1e18.
+    function _nomineeRelativeWeight(address nominee, uint256 chainId, uint256 time) internal view returns (uint256 weight) {
+        uint256 t = time / WEEK * WEEK;
+        uint256 totalSum = pointsSum[t].bias;
+
+        // Push a pair of key defining variables into one key
+        // nominee occupies first 160 bits
+        uint256 nomineeChainId = uint256(uint160(nominee));
+        // chain Id occupies no more than next 64 bits
+        nomineeChainId |= chainId << 160;
+
+        if (totalSum > 0) {
+            uint256 nomineeWeight = pointsWeight[nomineeChainId][t].bias;
+            weight = 1e18 * nomineeWeight / totalSum;
+        }
+    }
+
+    /// @dev Get Nominee relative weight (not more than 1.0) normalized to 1e18.
+    ///         (e.g. 1.0 == 1e18). Inflation which will be received by it is
+    ///         inflation_rate * relativeWeight / 1e18.
```

Comment:
In the original specs, if inflation_rate > votes_for_Nominee ->  votes_for_Nominee * relativeWeight / 1e18. Dispenser takes care of this?

Reply:
we will do it, using totalSum from this menthod